README.md

<!--  -->

<!-- README.md is generated from README.Rmd. Please edit that file -->

# vlad

<!-- [![Build Status](https://travis-ci.org/wittenberg/vlad.svg)](https://travis-ci.org/wittenberg/vlad) -->

<!-- [![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/vlad)](http://cran.r-project.org/package=vlad) -->

<!-- [![Coverage Status](https://codecov.io/gh/wittenberg/vlad/graph/badge.svg)](https://codecov.io/github/wittenberg/vlad?branch=master) -->

<!-- [![Downloads](https://cranlogs.r-pkg.org/badges/vlad)](https://CRAN.R-project.org/package=vlad) -->

<!-- [![Total Downloads](https://cranlogs.r-pkg.org/badges/grand-total/vlad?color=orange)](https://CRAN.R-project.org/package=vlad) -->

An R-package which contains functions to set up risk-adjusted quality
control charts in health care.

## Main features

  - Risk-adjusted CUSUM chart control limit calculations based on fast
    and accurate Markov chain approximations (Knoth, Wittenberg, and Gan
    [2019](#ref-Knoth.etal_2019))
  - Risk-adjusted CUSUM chart based on E-O (Wittenberg, Gan, and Knoth
    [2018](#ref-Wittenberg.etal_2018))
  - Risk-adjusted CUSUM chart based on log-likelihood ratio statistic
    (Steiner et al. [2000](#ref-Steiner.etal_2000))
  - Algorithms are implemented using Rcpp, RcppArmadillo and parallel
    computation

## Installation

You can install the released version of **vlad** from
[CRAN](https://CRAN.R-project.org/package=vlad) with:

``` r
install.packages("vlad")
```

You can install the development version from
[GitHub](https://github.com/wittenberg/vlad) with:

``` r
# install.packages("devtools")
devtools::install_github("wittenberg/vlad")
```

The package compiles C++ source code during installation, therefore you
need the appropriate compilers:

On *Windows* you need
[Rtools](https://cran.r-project.org/bin/windows/Rtools/) available from
CRAN.

On *macOS* you need the Clang 6.x compiler and the GNU Fortran compiler
from [macOS tools](https://cran.r-project.org/bin/macosx/tools/). Having
installed the compilers, you need to open a terminal and start R via
‘PATH=/usr/local/clang6/bin:$PATH R’. You can then install the package
via *devtools::install\_github(“wittenberg/vlad”)*

## Examples

[Supplemental material: Risk-adjusted CUSUM charts under model
error](https://onlinelibrary.wiley.com/action/downloadSupplement?doi=10.1002%2Fsim.8104&file=SIM_8104-Supp-0001-Supplemental-material-SIM-18-0571.R)

<!-- Load libraries: -->

<!-- ```{r load libraries, message=FALSE} -->

<!-- library("vlad") -->

<!-- library("dplyr") -->

<!-- library("tidyr") -->

<!-- library("ggplot2") -->

<!-- ``` -->

<!-- Subset the dataset `cardiacsurgery` into Phase I (first two years) and Phase II (five years) and estimate a risk model based on `phaseI`. -->

<!-- ```{r estimate risk model} -->

<!-- data("cardiacsurgery", package = "spcadjust") -->

<!-- cardiacsurgery <- cardiacsurgery %>% rename(s = Parsonnet) %>% -->

<!--   mutate(y = ifelse(status == 1 & time <= 30, 1, 0), -->

<!--         phase = factor(ifelse(date < 2*365, "I", "II"))) -->

<!-- head(cardiacsurgery) -->

<!-- phaseI <- filter(cardiacsurgery, phase == "I") %>% select(s, y) -->

<!-- coeff <- round(coef(glm(y ~ s, data = phaseI, family = "binomial")), 3) -->

<!-- print(coeff) -->

<!-- ``` -->

<!-- ### Create VLADs for seven surgeons -->

<!-- By using the estimated risk model coefficients `coeff`, for each pair of Parsonnet score `s` and operation outcome values `y`, the difference between expected and observed outcome is calculated with the function `calceo()`. -->

<!-- Thereafter, differences are cummulated to create the VLAD. This is done for all seven surgeons of the `cardiacsurgery` dataset. Results are saved to the object `vlads7`. -->

<!-- ```{r vlads7} -->

<!-- vlads7 <- lapply(1:7, function(j){ -->

<!--   Si <- filter(cardiacsurgery, surgeon == j) -->

<!--   EO <- sapply(seq_along(Si$s), function(i) calceo(df = Si[i, c("s", "y")], coeff = coeff)) -->

<!--   select(Si, surgeon, phase) %>%  mutate(n = 1:length(EO), cEO = cumsum(EO)) -->

<!-- })  -->

<!-- ``` -->

<!-- Create Variable life-adjusted Displays for each surgeon from the object `vlads7`. -->

<!-- ```{r VLADS1-7, fig.align='center', fig.width=8, fig.height=10} -->

<!-- vlads7 %>%  -->

<!--   bind_rows() %>%   -->

<!--   gather(key = "Surgeon", value = value, c(-n, -surgeon, -phase)) %>% -->

<!--   ggplot(aes(x = n, y = value, colour = phase, group = Surgeon)) + -->

<!--     geom_hline(yintercept = 0, colour = "darkgreen", linetype = "dashed") + -->

<!--     geom_line(size = 1.1) + facet_wrap( ~ surgeon, ncol = 2, scales = "free") + -->

<!--     labs(x="Patient number n", y="CUSUM E-O") + theme_classic() + -->

<!--     scale_y_continuous(sec.axis = dup_axis(name = NULL, labels = NULL)) + -->

<!--     scale_x_continuous(sec.axis = dup_axis(name = NULL, labels = NULL)) -->

<!-- ``` -->

<!-- ### Create a VLAD for surgeon 2 -->

<!-- ```{r vladS2, fig.align='center'} -->

<!-- S2 <- filter(cardiacsurgery, surgeon == 2) %>% select(phase, s, y) -->

<!-- S2I <- subset(S2, c(phase == "I")) -->

<!-- S2II <- subset(S2, c(phase == "II")) -->

<!-- coeff <- coef(glm(y ~ s, data = S2I, family = "binomial")) -->

<!-- EO <- sapply(1:nrow(S2), function(i) calceo(df = S2[i, c("s", "y")], coeff = coeff)) -->

<!-- df1 <- select(S2, phase) %>% mutate(n = row_number(), cEO = cumsum(EO)) -->

<!-- df2 <- gather(df1, variable, value, c(-n, -phase)) -->

<!-- p1 <- ggplot(df2, aes(x = n, y = value, colour = phase)) + -->

<!--   geom_hline(yintercept = 0, linetype = "dashed") + geom_line() + geom_point() +  -->

<!--   labs(x = "Patient number", y = "CUSUM E-O") + theme_classic() + -->

<!--   scale_y_continuous(sec.axis = dup_axis(name = NULL, labels = NULL)) + -->

<!--   scale_x_continuous(sec.axis = dup_axis(name = NULL, labels = NULL)) -->

<!-- p1 -->

<!-- ``` -->

<!-- ### Compute thresholds of a risk-adjusted CUSUM chart for surgeon 2  -->

<!-- Upper and lower control limits of the risk-adjusted CUSUM chart based on log-likelihood ratio statistic can be computed with the function `racusum_arl_h_sim()`. The implemention uses parallel simulation and a multi-stage search procedure.   -->

<!-- ```{r} -->

<!-- # set a random number generator for parallel computations -->

<!-- RNGkind("L'Ecuyer-CMRG") -->

<!-- # number of simulation runs -->

<!-- m <- 10^4 -->

<!-- # assign cores -->

<!-- nc <- parallel::detectCores() -->

<!-- # verbose calculation  -->

<!-- UCL_sim <- racusum_crit_sim(L0 = 740, df = S2I[, c("s", "y")], coeff = coeff, m = m, RA = 2, nc = nc, verbose = TRUE) -->

<!-- # quite calculation -->

<!-- LCL_sim <- racusum_crit_sim(L0 = 740, df = S2I[, c("s", "y")], coeff = coeff, m = m, RA = 1/2, nc = nc, verbose = FALSE) -->

<!-- round(cbind(UCL_sim, LCL_sim), 3) -->

<!-- ``` -->

### Authors

Philipp Wittenberg and Sven Knoth

### License

GPL (\>= 2)

### References

<div id="refs" class="references hanging-indent">

<div id="ref-Knoth.etal_2019">

Knoth, S, P Wittenberg, and FF Gan. 2019. “Risk-Adjusted CUSUM Charts
Under Model Error.” *Statistics in Medicine* 38 (12): 2206–18.
<https://doi.org/10.1002/sim.8104>.

</div>

<div id="ref-Steiner.etal_2000">

Steiner, SH, RJ Cook, VT Farewell, and T Treasure. 2000. “Monitoring
Surgical Performance Using Risk-Adjusted Cumulative Sum Charts.”
*Biostatistics* 1 (4): 441–52.
<https://doi.org/10.1093/biostatistics/1.4.441>.

</div>

<div id="ref-Wittenberg.etal_2018">

Wittenberg, P, FF Gan, and S Knoth. 2018. “A Simple Signaling Rule for
Variable Life-Adjusted Display Derived from an Equivalent Risk-Adjusted
CUSUM Chart.” *Statistics in Medicine* 37 (16): 2455–73.
<https://doi.org/10.1002/sim.7647>.

</div>

</div>


NEWS.md
# vlad 0.2.2
## Changes
* Combine documentation of multiple functions in same file

## Improvements
Fix ARL calculation for paired rounding approach for Markov Chain

# vlad 0.2.1
## Changes
* Use checkmate for error messages
* Calculate risk of death ahead of simulations to improve computation time
* Include various functions for distributions to compute the patient mix

## Improvements
* Improve simulation algorithms
* Improve testing

# vlad 0.2.0
## Changes
* Renamed functions:

original name         | new name
:---------------------|:-------------------
`cusum_arl_h_sim()`   | `bcusum_crit_sim()` 
`cusum_arl_sim()`     | `bcusum_arl_sim()`
`eocusum_arl_h_sim()` | `eocusum_crit_sim()`
`eocusum_adoc_sim()`  | `eocusum_ad_sim()`
`racusum_arl_h_sim()` | `racusum_crit_sim()`
`racusum_adoc_sim()`  | `racusum_ad_sim()`

## Improvements
* Include Markov Chain approximation for risk-adjusted CUSUM chart based on log-likelihood ratio statistic in function racusum_arl_mc() and racusum_crit_mc()
* Add option of calculation of optimal k based on mean outcome in function optimal_k()
* Add function for calculation of CUSUM scores for log-likelihood ratio CUSUM llr_cusum_scores()
and E-O CUSUM eo_cusum_scores()
* Extend unit tests for functions

# vlad 0.1.0
## Improvements
* Preparation for CRAN release


R/racusum_crit.R
#' @name racusum_crit
#' @title Alarm thresholds of RA-CUSUM charts
#' @description Compute alarm threshold of risk-adjusted CUSUM charts.
#' @param L0 Double. Prespecified Average Run Length.
#' @param jmax Integer. Number of digits for grid search.
#' @param pmix Numeric Matrix. A three column matrix. First column is the risk
#' score distribution. Second column are the predicted probabilities from the risk model. Third
#'  column can be either the predicted probabilities from the risk model or average outcome per
#'  risk score, see examples.
#' @param RA Double. Odds ratio of death under the alternative hypotheses. Detecting deterioration
#' in performance with increased mortality risk by doubling the odds Ratio \code{RA = 2}. Detecting
#'  improvement in performance with decreased mortality risk by halving the odds ratio of death
#'  \code{RA = 1/2}. Odds ratio of death under the null hypotheses is \code{1}.
#' \code{RQ}. Use \code{RQ = 1} to compute the in-control ARL and other values to compute the
#' out-of-control ARL.
#' @param RQ Double. Defines the true performance of a surgeon with the odds ratio ratio of death
#' \code{RQ}. Use \code{RQ = 1} to compute the in-control ARL and other values to compute the
#' out-of-control ARL.
#' @param scaling Double. The \code{scaling} parameter controls the quality of the approximation,
#' larger values achieve higher accuracy but increase the computation burden (larger transition
#' probability matrix).
#' @param rounding Character. If \code{rounding = "p"} a paired rounding implementation of
#' \emph{Knoth et al. (2019)} is used, if \code{rounding = "s"} a simple rounding method of
#' \emph{Steiner et al. (2000)} is used.
#' @param method Character. If \code{method = "Toep"} a combination of Sequential Probability Ratio
#'  Test and Toeplitz matrix structure is used to calculate the ARL. \code{"ToepInv"} computes the
#'  inverted matrix using Toeplitz matrix structure. \code{"BE"} solves a linear equation system
#'  using the classical approach of \emph{Brook and Evans (1972)} to calculate the ARL.
#' @param verbose Logical. If \code{FALSE} a quiet calculation of \code{h} is done. If \code{TRUE}
#'  verbose output of the search procedure is included.
#' @return Returns a single value which is the control limit \code{h} for a given In-control ARL.
#' @details
#' Determines the control limit for given in-control ARL (\code{"L0"}) using
#' \code{\link{racusum_arl_mc}} by applying a grid search.
#' @author Philipp Wittenberg
#' @references Knoth S, Wittenberg P and Gan FF (2019).
#' Risk-adjusted CUSUM charts under model error.
#' \emph{Statistics in Medicine}, \strong{38}(12), pp. 2206--2218.
#'
#' Wittenberg P, Gan FF, Knoth S (2018).
#' A simple signaling rule for variable life-adjusted display derived from
#' an equivalent risk-adjusted CUSUM chart.
#' \emph{Statistics in Medicine}, \strong{37}(16), pp 2455--2473.
#'
#' Steiner SH, Cook RJ, Farewell VT and Treasure T (2000).
#'  Monitoring surgical performance using risk-adjusted cumulative sum charts.
#'  \emph{Biostatistics}, \strong{1}(4), pp. 441--452.
#'
#' Brook D and Evans DA (1972)
#'  An approach to the probability distribution of CUSUM run length.
#'  \emph{Biometrika}, \strong{59}(3), pp. 539--549
#'
#' Webster RA and Pettitt AN (2007)
#' Stability of approximations of average run length of risk-adjusted CUSUM schemes using
#' the Markov approach: comparing two methods of calculating transition probabilities.
#'  \emph{Communications in Statistics - Simulation and Computation} \strong{36}(3), pp. 471--482
#' @examples
#' \dontrun{
#' library(vlad)
#' library(dplyr)
#' data("cardiacsurgery", package = "spcadjust")
#'
#' ## Markov Chain
#' ## preprocess data to 30 day mortality and subset phase I (In-control) of surgeons 2
#' S2I <- cardiacsurgery %>% rename(s = Parsonnet) %>%
#'   mutate(y = ifelse(status == 1 & time <= 30, 1, 0),
#'          phase = factor(ifelse(date < 2*365, "I", "II"))) %>%
#'   filter(phase == "I", surgeon == 2) %>% select(s, y)
#'
#' ## estimate risk model, get relative frequences and probabilities
#' mod1 <- glm(y ~ s, data = S2I, family = "binomial")
#' fi  <- as.numeric(table(S2I$s) / length(S2I$s))
#' usi <- sort(unique(S2I$s))
#' pi1 <- predict(mod1, newdata = data.frame(s = usi), type = "response")
#'
#' ## set up patient mix
#' pmix  <- data.frame(fi, pi1, pi1)
#'
#' ## control limit for detecting deterioration RA = 2:
#' racusum_crit_mc(pmix = pmix, L0 = 740, RA = 2, RQ = 1)
#' ## control limit for detecting improvement RA = 1/2:
#' racusum_crit_mc(pmix = pmix, L0 = 740, RA = 0.5, RQ = 1)
#'
#' ## Monte Carlo simulation
#' SALL <- cardiacsurgery %>% rename(s = Parsonnet) %>%
#'   mutate(y = ifelse(status == 1 & time <= 30, 1, 0),
#'          phase = factor(ifelse(date < 2*365, "I", "II")))
#' SI <- subset(SALL, phase == "I")
#' y <- subset(SALL, select = y)
#' GLM <- glm(y ~ s, data = SI, family = "binomial")
#' pi1 <- predict(GLM, type = "response", newdata = data.frame(s = SALL$s))
#' pmix <- data.frame(y, pi1, pi1)
#' h <- racusum_crit_sim(pmix = pmix, L0 = 370, RA = 2, nc = 4, verbose = TRUE)
#' }
#' @export
racusum_crit_mc <- function(L0, pmix, RA, RQ, scaling = 600, rounding = "p", method = "Toep", jmax = 4, verbose = FALSE) {
  arg_checks <- checkmate::makeAssertCollection()
  pmix <- as.matrix(pmix)
  checkmate::assert_matrix(pmix, ncols = 3, add = arg_checks)
  checkmate::assert_integerish(L0, lower = 0, add = arg_checks)
  checkmate::assert_numeric(RA, len = 1, lower = 0, add = arg_checks)
  checkmate::assert_numeric(RQ, len = 1, lower = 0, add = arg_checks)
  checkmate::assert_integerish(scaling, lower = 1, add = arg_checks)
  rounding <- tolower(rounding)
  checkmate::assert_choice(rounding, c("p", "s"))
  irounding <- switch(rounding, p = 1, s = 2)
  checkmate::assert_choice(method, c("Toep", "ToepInv", "BE"))
  imethod <- switch(method, Toep = 1, ToepInv = 2, BE = 3)
  checkmate::assert_integerish(jmax, lower = 0, add = arg_checks)
  checkmate::assert_logical(verbose, len = 1, add = arg_checks)
  if (!arg_checks$isEmpty()) checkmate::reportAssertions(arg_checks)
  .racusum_crit_mc(pmix, L0, RA, RQ, scaling, irounding, imethod, jmax, verbose)
}

#' @rdname racusum_crit
#' @param yemp Logical. If \code{TRUE}, use emirical outcome values, else use model.
#' @param m Integer. Number of simulation runs.
#' @param nc Integer. Number of cores used for parallel processing. Value is passed to
#' \code{\link{parSapply}}.
#' @param hmax Integer. Maximum value of \code{h} for the grid search.
#' @details Determines the control limit ("\code{h}") for given in-control ARL (\code{"L0"})
#' applying a grid search using \code{\link{racusum_arl_sim}} and \code{\link{parSapply}}.
#' @export
racusum_crit_sim <- function(L0, pmix, RA = 2, RQ = 1, yemp = FALSE, m = 1e4, nc = 1, hmax = 30, jmax = 4, verbose = FALSE) {
  arg_checks <- checkmate::makeAssertCollection()
  checkmate::assert_numeric(L0, len = 1, lower = 0, add = arg_checks)
  checkmate::assert_data_frame(pmix, ncols = 3, add = arg_checks)
  checkmate::assert_numeric(RA, len = 1, lower = 0, add = arg_checks)
  checkmate::assert_numeric(RQ, len = 1, lower = 0, add = arg_checks)
  checkmate::assert_logical(yemp, len = 1, add = arg_checks)
  checkmate::assert_integerish(m, lower = 1, add = arg_checks)
  checkmate::assert_integerish(nc, lower = 1, add = arg_checks)
  checkmate::assert_integerish(hmax, len = 1, lower = 1, add = arg_checks)
  checkmate::assert_integerish(jmax, len = 1, lower = 1, add = arg_checks)
  checkmate::assert_logical(verbose, len = 1, add = arg_checks)
  if (!arg_checks$isEmpty()) checkmate::reportAssertions(arg_checks)

  cl <- parallel::makeCluster(getOption("cl.cores", nc))

  for ( h in 1:hmax ) {
    parallel::clusterExport(cl, c("h", "racusum_arl_sim", "pmix", "RA", "RQ", "yemp", "m"), envir = environment())
    L1 <- mean(parallel::parSapply(cl, 1:m, racusum_arl_sim, pmix = pmix, h = h, RA = RA, RQ = RQ, yemp = yemp))
    if ( verbose ) cat(paste("h =", h, "\tARL =", L1, "\n"))
    if ( L1 > L0 ) break
  }
  h1 <- h

  for ( j in 1:jmax ) {
    for ( dh in 1:19 ) {
      h <- h1 + (-1)^j*dh/10^j
      parallel::clusterExport(cl, c("h", "racusum_arl_sim", "pmix", "RA", "RQ", "yemp", "m"), envir = environment())
      L1 <- mean(parallel::parSapply(cl, 1:m, racusum_arl_sim, pmix = pmix, h = h, RA = RA, RQ = RQ, yemp = yemp))
      if ( verbose ) cat(paste("h =", h, "\tARL =", L1, "\n"))
      if ( (j %% 2 == 1 & L1 < L0) | (j %% 2 == 0 & L1 > L0) ) break
    }
    h1 <- h
  }

  if ( L1 < L0 ) h <- h + 1/10^jmax
  parallel::stopCluster(cl)
  h
}

